### PR TITLE
Detect if OpenShift cluster is single stack ipv6

### PR DESCRIPTION
**Pull request by dharmit**
**What this PR does / why we need it**:
Checks if HCO is running on OCP configured with single stack ipv6 setup, and fail starting the Pod if so.

**Reviewer Checklist**

> Reviewers are supposed to review the PR for every aspect below one by one. To check an item means the PR is either "OK" or "Not Applicable" in terms of that item. All items are supposed to be checked before merging a PR. 

- [ ] PR Message
- [ ] Commit Messages
- [ ] How to test
- [ ] Unit Tests
- [ ] Functional Tests
- [ ] User Documentation
- [ ] Developer Documentation
- [ ] Upgrade Scenario
- [ ] Uninstallation Scenario
- [ ] Backward Compatibility
- [ ] Troubleshooting Friendly

**Jira Ticket**:

```jira-ticket
https://issues.redhat.com/browse/CNV-28584
```

**Release note**:

```release-note
Detect and alert if OCP is single stack IPv6
```

**NOTE**:
We need to get https://github.com/kubevirt/monitoring/pull/186 merged before opening this up for review because it will keep failing the following functional test in `monitoring_test.go` otherwise:
```go
It("Alert rules should have all the requried annotations", func() {
```

### CODE REVIEW

File: pkg/util/cluster.go [hunk 1]
```diff
@@ -142,6 +148,24 @@ func (c *ClusterInfoImp) initOpenshift(ctx context.Context, cl client.Client) er
...
+		return err
+	}
+	cn := clusterNetwork.Status.ClusterNetwork
+	for _, i := range cn {
+		c.logger.Info("Cluster Network",
+			"CIDR", i.CIDR,
+			"Host Prefix", i.HostPrefix,
+		)
+	}
+	c.singlestackipv6 = len(cn) == 1 && strings.Contains(cn[0].CIDR, ":")
```
Comment (dharmit) [inline pkg/util/cluster.go:168]: I'm not 100% sure if this is the best way to detect if the OCP cluster is configured as single stack IPv6, but it's my best guess after reading https://docs.openshift.com/container-platform/4.13/networking/ovn_kubernetes_network_provider/converting-to-dual-stack.html.

Open to other/better/canonical approaches!

File: controllers/alerts/alerts.go [hunk 2]
```diff
@@ -27,6 +27,7 @@ const (
 	outOfBandUpdateAlert          = "KubevirtHyperconvergedClusterOperatorCRModification"
 	unsafeModificationAlert       = "KubevirtHyperconvergedClusterOperatorUSModification"
 	installationNotCompletedAlert = "KubevirtHyperconvergedClusterOperatorInstallationNotCompletedAlert"
+	singleStackIPv6Alert          = "KubevirtHyperconvergedClusterOperatorSingleStackIPv6Alert"
```
Comment (machadovilaca) [inline controllers/alerts/alerts.go:30]: I see `KubevirtHyperconvergedClusterOperatorInstallationNotCompletedAlert `, but the alerts should not be suffixed with the word `Alert`

File: controllers/alerts/alerts.go [hunk 3]
```diff
@@ -215,3 +217,18 @@ func createOperatorHealthStatusRule() monitoringv1.Rule {
...
+func createSingleStackIPv6AlertRule() monitoringv1.Rule {
+	return monitoringv1.Rule{
+		Alert: singleStackIPv6Alert,
+		Expr:  intstr.FromString("kubevirt_hco_hyperconverged_single_stack_ipv6 == 0"),
+		Annotations: map[string]string{
+			"description": "OpenShift Virtualization is not supported on a single stack IPv6 cluster",
+			"summary":     "OpenShift Virtualization is not supported on a single stack IPv6 cluster",
+		},
+		Labels: map[string]string{
+			severityAlertLabelKey:     "warning",
```
Comment (machadovilaca) [inline controllers/alerts/alerts.go:230]: this prevents users from running the project, right? if so, it should be "critical"

File: controllers/alerts/alerts.go [hunk 4]
```diff
@@ -215,3 +217,18 @@ func createOperatorHealthStatusRule() monitoringv1.Rule {
 		Expr:   intstr.FromString(`label_replace(vector(2) and on() ((kubevirt_hco_system_health_status>1) or (count(ALERTS{kubernetes_operator_part_of="kubevirt", alertstate="firing", operator_health_impact="critical"})>0)) or (vector(1) and on() ((kubevirt_hco_system_health_status==1) or (count(ALERTS{kubernetes_operator_part_of="kubevirt", alertstate="firing", operator_health_impact="warning"})>0))) or vector(0),"name","kubevirt-hyperconverged","","")`),
 	}
 }
+
+func createSingleStackIPv6AlertRule() monitoringv1.Rule {
+	return monitoringv1.Rule{
+		Alert: singleStackIPv6Alert,
+		Expr:  intstr.FromString("kubevirt_hco_hyperconverged_single_stack_ipv6 == 0"),
```
Comment (machadovilaca) [inline controllers/alerts/alerts.go:224]: > 	SingleStackIPv6True  = float64(1)
	SingleStackIPv6False = float64(0)
	
shouldn't this be `kubevirt_hco_hyperconverged_single_stack_ipv6 == 1`?

**Review by machadovilaca — CHANGES_REQUESTED** [review]
please also add a functional test for the alert

File: controllers/alerts/alerts.go [hunk 5]
```diff
@@ -215,3 +217,18 @@ func createOperatorHealthStatusRule() monitoringv1.Rule {
 		Expr:   intstr.FromString(`label_replace(vector(2) and on() ((kubevirt_hco_system_health_status>1) or (count(ALERTS{kubernetes_operator_part_of="kubevirt", alertstate="firing", operator_health_impact="critical"})>0)) or (vector(1) and on() ((kubevirt_hco_system_health_status==1) or (count(ALERTS{kubernetes_operator_part_of="kubevirt", alertstate="firing", operator_health_impact="warning"})>0))) or vector(0),"name","kubevirt-hyperconverged","","")`),
 	}
 }
+
+func createSingleStackIPv6AlertRule() monitoringv1.Rule {
+	return monitoringv1.Rule{
+		Alert: singleStackIPv6Alert,
+		Expr:  intstr.FromString("kubevirt_hco_hyperconverged_single_stack_ipv6 == 1"),
```
Comment (sradco) [inline controllers/alerts/alerts.go:224]: @dharmit  I think the metric name is incorrect here. Should be kubevirt_hco_single_stack_ipv6

Comment (dharmit) [reply controllers/alerts/alerts.go:224]: Sure. I simply took the cue from another alert defined a few lines earlier in the same file:
```go
Expr:  intstr.FromString("kubevirt_hco_hyperconverged_cr_exists == 0"),
```

Comment (dharmit) [reply controllers/alerts/alerts.go]: I think the problem was I had used `kubevirt_hco_single_stack_ipv6` in `metrics.go`, but `kubevirt_hco_hyperconverged_single_stack_ipv6` in `alerts.go`. I have modified as you suggested.

File: controllers/alerts/alerts.go [hunk 6]
```diff
@@ -27,6 +27,7 @@ const (
 	outOfBandUpdateAlert          = "KubevirtHyperconvergedClusterOperatorCRModification"
 	unsafeModificationAlert       = "KubevirtHyperconvergedClusterOperatorUSModification"
 	installationNotCompletedAlert = "KubevirtHyperconvergedClusterOperatorInstallationNotCompletedAlert"
+	singleStackIPv6Alert          = "KubevirtHCOSingleStackIPv6"
```
Comment (machadovilaca) [inline controllers/alerts/alerts.go:30]: based on https://github.com/kubevirt/monitoring/pull/188, please rename the alert name to `KubeVirtHcoSingleStackIPv6`

**Review by machadovilaca — CHANGES_REQUESTED** [review]

File: controllers/alerts/alerts.go (hunk 6 above)
Comment (dharmit) [reply controllers/alerts/alerts.go]: Done.

**Comment by tiraboschi** [comment]
This is currently failing CI on:
```
  [FAILED] KubeVirtHcoSingleStackIPv6 runbook is not available
  Expected
      <int>: 404
  to equal
      <int>: 200
  In [It] at: /go/src/github.com/kubevirt/hyperconverged-cluster-operator/tests/func-tests/monitoring_test.go:73 @ 07/25/23 08:39:05.421
  ```
  can you please take a look?

**Comment by dharmit** [comment]
> This is currently failing CI on:
> 
> ```
>   [FAILED] KubeVirtHcoSingleStackIPv6 runbook is not available
>   Expected
>       <int>: 404
>   to equal
>       <int>: 200
>   In [It] at: /go/src/github.com/kubevirt/hyperconverged-cluster-operator/tests/func-tests/monitoring_test.go:73 @ 07/25/23 08:39:05.421
> ```
> 
> can you please take a look?

Now that https://github.com/kubevirt/monitoring/pull/188 is merged, I can address the issue here. :)

File: tests/func-tests/utils.go [hunk 7]
```diff
@@ -109,12 +119,46 @@ func (c *cacheIsOpenShift) IsOpenShift(cli kubecli.KubevirtClient) (bool, error)
...
+	s := scheme.Scheme
+	_ = openshiftconfigv1.Install(s)
+	s.AddKnownTypes(openshiftconfigv1.GroupVersion)
+
+	clusterNetwork := &openshiftconfigv1.Network{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "cluster",
+		},
+	}
+	err = cli.RestClient().Get().
```
Comment (tiraboschi) [inline tests/func-tests/utils.go:138]: can you please use `cli.DynamicClient().Resource(GVR)...`?

**Pull request merged**